src/main/java/org/jabref/gui/theme/ThemeImpl.java

```diff
@@ -0,0 +1,219 @@
+package org.jabref.gui.theme;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.WeakHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import javafx.scene.Scene;
+import javafx.scene.web.WebEngine;
+
+import org.jabref.gui.util.Theme;
+import org.jabref.model.util.FileUpdateMonitor;
+import org.jabref.preferences.AppearancePreferences;
+import org.jabref.preferences.PreferencesService;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Installs the style file and provides live reloading.
+ * JabRef provides two inbuilt themes and a user customizable one: Light, Dark and Custom. The Light theme is basically
+ * the base.css theme. Every other theme is loaded as an addition to base.css.
+ *
+ * For type Custom, Theme will protect against removal of the CSS file, degrading as gracefully as possible. If the file
+ * becomes unavailable while the application is running, some Scenes that have not yet had the CSS installed may not be
+ * themed. The PreviewViewer, which uses WebEngine, supports data URLs and so generally are not affected by removal
+ * of the file; however Theme will not attempt to URL-encode large style sheets so as to protect
+ * memory usage (see {@link StyleSheetFile#MAX_IN_MEMORY_CSS_LENGTH}.
+ *
+ * @see <a href="https://docs.jabref.org/advanced/custom-themes">Custom themes</a> in the Jabref documentation.
+ */
+public class ThemeImpl implements Theme {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ThemeImpl.class);
+
+    private final PreferencesService preferencesService;
+
+    private final FileUpdateMonitor fileUpdateMonitor;
+
+    private final Consumer<Runnable> updateRunner;
+
+    private final StyleSheet baseStyleSheet;
+
+    private final AtomicReference<ThemeState> state = new AtomicReference<>();
+
+    private final Set<Scene> scenes = Collections.newSetFromMap(new WeakHashMap<>());
+    private final Set<WebEngine> webEngines = Collections.newSetFromMap(new WeakHashMap<>());
+
+    private static final class ThemeState {
+
+        private final ThemePreference themePreference;
+
+        private final StyleSheet additionalStylesheet;
```

Comment:
Maybe make this an `Optional<StyleSheet>`? Than you would not need to introduce the "EmptyStylesheet" class.

Reply:
The idea here is to have an invariant: "there is always an additional stylesheet".

This is so that it is much simpler to transition between states where an additional stylesheet is available, and states where it is not (either because it is missing, or because the theme does not require one).

Reply:
ok, but "there is always an additional stylesheet" doesn't seem to be valid from a business logic (the light theme doesn't have additional css) as well as from the current implementation (EmptyStylesheet is not a a stylesheet, it's something that pretends to be one).

On the other hand, the current solution works and is fine codewise so take my comment as a suggestion that doesn't necessarily need to be implemented.

Reply:
Ah, perhaps it wasn't clear! EmptyStylesheet loaded a real resource called Empty.css so it wasn't pretending!

However, given it is confusing, I've updated to remove StyleSheetEmpty and rename Empty.css to Light.css, which has a comment in it explaining that it doesn't have to add anything over Bas.css

This is meaningful, because you _could_ add something specific to Light.css. A practical example would be theming for Preview. That can't be done in Base.css, because WebEngine (which preview uses) can only take one style sheet and that is the additional one. So Light.css would be the place for that.

So hopefully that clears things up while keeping the useful invariant.